File: TerraObserver/Godot.Abstractions/Extensions/Chunks/IHexMesh.cs
using Godot.Abstractions.Bases;

namespace Godot.Abstractions.Extensions.Chunks;

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-06-08 15:22:08
public interface IHexMesh : IMeshInstance3D
{
    void Clear();
    void Apply();
    void ShowMesh(Mesh mesh);
    void AddTriangle(Vector3[] vs, Color[]? tws = null,
        Vector2[]? uvs = null, Vector2[]? uvs2 = null, Vector3 tis = default);
    void AddQuad(Vector3[] vs, Color[]? tws = null,
        Vector2[]? uvs = null, Vector2[]? uvs2 = null, Vector3 tis = default);
}

File: TerraObserver/TO.Domains/Types/Chunks/IHexMesh.fs
namespace TO.Domains.Types.Chunks

open System.Runtime.InteropServices
open Godot
open TO.Domains.Types.Godots

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-06-29 13:38:29
[<Interface>]
type IHexMesh =
    inherit IMeshInstance3D
    // =====【Export】=====
    abstract UseCollider: bool with get
    abstract UseCellData: bool with get
    abstract UseUvCoordinates: bool with get
    abstract UseUv2Coordinates: bool with get
    abstract Smooth: bool with get
    // =====【普通属性】=====
    abstract SurfaceTool: SurfaceTool with get, set
    abstract VIdx: int with get, set

    abstract AddTriangle:
        vs: Vector3 array *
        [<Optional; DefaultParameterValue(null: Color array)>] tws: Color array *
        [<Optional; DefaultParameterValue(null: Vector2 array)>] uvs: Vector2 array *
        [<Optional; DefaultParameterValue(null: Vector2 array)>] uvs2: Vector2 array *
        [<Optional; DefaultParameterValue(Vector3())>] tis: Vector3 ->
            unit

    abstract AddQuad:
        vs: Vector3 array *
        [<Optional; DefaultParameterValue(null: Color array)>] tws: Color array *
        [<Optional; DefaultParameterValue(null: Vector2 array)>] uvs: Vector2 array *
        [<Optional; DefaultParameterValue(null: Vector2 array)>] uvs2: Vector2 array *
        [<Optional; DefaultParameterValue(Vector3())>] tis: Vector3 ->
            unit
